src/collapse/perf.rs

```diff
@@ -88,28 +72,54 @@ struct PerfState {
     /// Called pname after original stackcollapse-perf source.
     pname: String,
 
-    /// The options for the current run.
-    opt: Options,
-
     event_filtering: EventFilterState,
+
+    opt: PerfOptions,
+}
+
+impl Frontend for Perf {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.on_line(line.trim_end());
+            }
+        }
+        self.finish(writer)
+    }
 }
 
-impl From<Options> for PerfState {
-    fn from(opt: Options) -> Self {
-        PerfState {
+impl Perf {
+    pub fn new(options: PerfOptions) -> Self {
```

Comment:
```suggestion
    pub fn from_options(options: PerfOptions) -> Self {
```

Reply:
👍

Reply:
And while we're at it, maybe rename the argument to just `opt` so you can use it directly in the constructor below.

Reply:
👍